client/src/ScanPage/AddressTokensPage.tsx

```diff
@@ -24,94 +26,111 @@ type AddressTokensPageState = {
     tokens: null,
     error: false,
     address: null,
     ens: null,
     loading: false,
+    isRedeemModalOpen: false,
+    isRedeemLoading: false,
+  });
+
+  const { addToast } = useToasts();
+
+  const { tokens, error, address, ens, loading, isRedeemLoading, isRedeemModalOpen } = state;
+
+  useEffect(() => {
+    getTokens();
+  }, []); // eslint-disable-line 
+
+  const handleOpenReedemModalClick = () => {
+    setState((oldState) => ({ ...oldState, isRedeemModalOpen: true }));
   };
 
-  componentDidMount() {
-    this.getTokens();
-  }
+  const handleCloseReedemModalClick = () => {
+    setState((oldState) => ({ ...oldState, isRedeemModalOpen: false }));
+  };
 
-  async getTokens() {
-    try {
-      this.setState({ loading: true });
+  const handleRedeemConfirm = () => {
+    setState((oldState) => ({ ...oldState, isRedeemLoading: true }));
 
-      const { location, match } = this.props;
-      const account = delve(match, 'params.account');
-      const addressFromHistory = delve(location, 'state.address');
-      const address = account || addressFromHistory;
+    redeemPoaps()
+      .then(() => {
+        setState((oldState) => ({ ...oldState, isRedeemModalOpen: false }));
 
-      if (isValidAddress(address)) {
-        const tokens = await getTokensFor(address);
-        const ens = await getENSFromAddress(address);
+        const successMessage = 'Your Poaps were redeemed successfully';
 
-        this.setState({ tokens, address, ens: ens.valid ? ens.ens : null });
-      } else {
-        const ensResponse = await resolveENS(address);
+        addToast(successMessage, {
+          appearance: 'success',
+          autoDismiss: true,
+        });
+      })
+      .catch(() => {
+        const errorMessage = 'An error occurred redeeming your Poaps';
 
-        if (ensResponse.valid) {
-          const tokens = await getTokensFor(ensResponse.ens);
-          this.setState({ tokens, address: ensResponse.ens, ens: address });
-        }
-      }
-    } catch (err) {
-      this.setState({ error: true });
-    } finally {
-      this.setState({ loading: false });
-    }
-  }
+        addToast(errorMessage, {
+          appearance: 'error',
+          autoDismiss: false,
+        });
+      })
+      .finally(() => setState((oldState) => ({ ...oldState, isRedeemLoading: false })));
+  };
 
-  getTokensByYear(): {
+  type Bla = {
     year: number;
     tokens: TokenInfo[];
-  }[] {
-    if (this.state.tokens == null) {
+  };
+
+  const getTokensByYear = (): Bla[] => {
+    if (state.tokens == null) {
       throw new Error('There are no tokens');
     }
+
     const tokensByYear: Map<number, TokenInfo[]> = new Map();
-    for (const t of this.state.tokens) {
+
+    for (const t of state.tokens) {
       if (tokensByYear.has(t.event.year)) {
         tokensByYear.get(t.event.year)!.push(t);
       } else {
         tokensByYear.set(t.event.year, [t]);
       }
     }
```

Comment:
Avoid short abbreviated names and prefer destructuring

```suggestion
    for (const token of state.tokens) {
      const { year } = token.event;
      
      if (tokensByYear.has(year)) {
        tokensByYear.get(year)!.push(t);
      } else {
        tokensByYear.set(year, [t]);
      }
    }
```